Add OIDC authentication support with multi-provider configuration

AlexBocken opened the pull request:
Adds comprehensive OIDC authentication support to Gramps Web API with multi-provider configuration and automatic role mapping.

## Features

  - Multi-provider support: Google, Microsoft, GitHub, and custom OIDC providers
  - Automatic role mapping: Map OIDC groups/claims to Gramps user roles
  - Flexible configuration: Support for various claim structures and nested groups
  - Backward compatibility: Preserves existing authentication when OIDC is disabled
  - Security: Proper token validation and user management

## Configuration

  All OIDC settings use GRAMPSWEB_OIDC_* environment variables for consistency. Role mapping automatically assigns permissions based on group membership from OIDC tokens.

 ## Testing

 Complete testing environment available at: https://github.com/AlexBocken/gramps-oidc-testing
 
 There are some unit-tests included but to be honest, these still need improving.
 
 ## Misc
 
 Happy to improve the codebase based on feedback, I tried to stick to utility functions and similar coding style to not waste effort.

 **Related**: Requires companion frontend PR for full OIDC integration. (PR [gramps-web/#763)(https://github.com/gramps-project/gramps-web/pull/763)

DavidMStraub commented:
Thanks a lot! :partying_face:  Will review ASAP.

Fixes #355

AlexBocken commented:
Noticed some `os.getenv` were able to get through (using the default names in configs is not helpful for debugging...)

AlexBocken commented:
Further investigation with OIDC Provider alternatives to keycloak revealed some issues. Authentik does not seem to want to cooperate in the token exchange with the current setup. Investigating.

DavidMStraub commented on README.md:
```diff
@@ -12,6 +12,32 @@ Gramps Web API is the backend of [Gramps Web](https://www.grampsweb.org/), a gen
 - Developer documentation for Gramps Web API: https://www.grampsweb.org/dev-backend/
 - Documentation for Gramps Web: https://www.grampsweb.org
 
+## OIDC Configuration
```
This is not the right place to document configuration. Please submit a PR to https://github.com/gramps-project/gramps-web-docs.

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,202 @@
+#
+# Gramps Web API - A RESTful API for the Gramps genealogy program
+#
+# Copyright (C) 2020-2024      David Straub
```
I guess you want to update that header in all files you added

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,202 @@
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""OIDC authentication endpoint blueprint."""
```
This is not a blueprint

AlexBocken replied:
adjusted to "resources" instead of  "endpoint blueprint". Follows a similar pattern to other files in the repo.

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,202 @@
+    def get(self, args):
+        """Redirect to OIDC provider for authentication."""
+        if not is_oidc_enabled():
+            abort_with_message(404, "OIDC authentication is not enabled")
```
I would return 405 here

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,202 @@
+    def get(self, args):
+        """Handle OIDC callback and create JWT tokens."""
+        if not is_oidc_enabled():
+            abort_with_message(404, "OIDC authentication is not enabled")
```
405

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,202 @@
+                userinfo = oidc_client.userinfo(token=token)
+
+        except Exception as e:
+            logger.error(f"OIDC callback error for provider '{provider_id}': {e}")
```
What about

```python
logger.exception(f"OIDC callback error for provider '{provider_id}'")
```

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,360 @@
+
+    # For GitHub, get email from separate API call if not in userinfo
+    if provider_id == "github" and not email:
+        email = userinfo.get("email") or f"{userinfo.get('login', 'user')}@users.noreply.github.com"
```
Will emails to @users.noreply.github.com be forwarded to the user?

AlexBocken replied:
Good catch. No, they won't. Can only use this email for signing of commits on GitHub, my mistake. Removing this special case then.

DavidMStraub commented on gramps_webapi/auth/oidc.py:
```diff
@@ -0,0 +1,360 @@
+    # Check if any role mapping is configured
+    has_role_mapping = any(group_name.strip() for group_name in role_mapping.values())
+    if not has_role_mapping:
+        logger.info("No OIDC role mapping configured (no GRAMPSWEB_OIDC_GROUP_* environment variables set). Preserving existing user roles.")
```
Not environment variables but generally config options (without `GRAMPSWEB_`)

AlexBocken commented:
Alright, I have addressed all your raised points. Besides this the following items are imo still open and will be addressed by me in the coming days:

1. proper logout handling for OIDC (otherwise we do not log out on OIDC provider side if we click "Logout" in gramps-web)
2. unit-testing cleanup and full coverage
3. Test with other OIDC providers (e.g. Authentik) and fix any problems

DavidMStraub commented on .devcontainer/docker-compose.yml:
```diff
@@ -27,6 +27,7 @@ services:
       GRAMPSWEB_CELERY_CONFIG__broker_url: redis://redis:6379/0
       GRAMPSWEB_CELERY_CONFIG__result_backend: redis://redis:6379/0
       GRAMPSWEB_LOG_LEVEL: DEBUG
+      GRAMPSWEB_FRONTEND_URL: http://localhost:8001
```
Let's make this empty by default rather than the development value

DavidMStraub commented:
Thanks for the changes and the explanations!

Just one more issue I noticed: defaulting to `ROLE_GUEST` means the entire Gramps Web instance (except private records) is opened up to everyone on the internet with a Google account. This cannot be the default! IMO we should default to `ROLE_DISABLED` and dispatch `send_email_new_user` exactly like in the registration endpoint.

DavidMStraub commented on gramps_webapi/auth/__init__.py:
```diff
@@ -201,6 +202,16 @@ def _get_user_detail(user, include_guid: bool = False):
+        details["oidc_accounts"] = oidc_accounts
+        # Add a simplified account source summary for frontend display
+        if oidc_accounts:
+            from ..api.util import get_config
```
Let's import this at the top please

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,281 @@
+            )
+
+            # Redirect to frontend with secure HTTP-only cookies
+            from flask import redirect, make_response
```
Top-level import please

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,281 @@
+    def get(self):
+        """Exchange HTTP-only cookies for tokens that can be stored in localStorage."""
+        from flask import request as flask_request, make_response, jsonify
+        import os
```
All top level

DavidMStraub commented on gramps_webapi/api/resources/oidc.py:
```diff
@@ -0,0 +1,281 @@
+            'access_token': access_token,
+            'refresh_token': refresh_token,
+            'token_type': 'Bearer'
+        }))
```
I'm pretty sure you can omit `make_response` *and* `jsonify` in modern Flask, that makes this more readable (also valid in other places)

AlexBocken replied:
Good catch. Removed the make_response as it really is unnecessary but kept the jsonify. Without the jsonify the cookie setting becomes too unreadable imo.

DavidMStraub commented:
Great work, thanks for your efforts, I think we are close.

But I think I found one more security issue.

If a user exists with email address A and tree X, they can sing up for tree Y with an OIDC account that has the same email address, and the code will update the existing user to use the new tree, but preserving the role. So, say, any tree owner on a given multi tree server could escalate to become tree owner of any other tree.

Although we could close this whole with appropriate checks, I think we should remove the email matching feature from the scope of this PR to keep it simpler. Can you please remove it? We can always add it back later. Sorry for suggesting that initially.

Apart from that, two questions:

1. Would it be easy to add an endpoint to associate an *existing* user account with an OIDC account manually? I'm thinking of something like a "associate Google login" button in the user profile. Again this would be out of scope of the present PR but I want to make sure the architecture can be modified to allow it in the future (without more user db schema changes)
2. What is the rationale for the `last_login`? I know it's "good to know", but we don't track it for "normal" user accounts either. Is it needed for any security or functional purpose? If not, I suggest removing it.

AlexBocken commented:
Oh wow, good catch with the email matching privilege escalation! Apologies, I did not consider the multi-tree scenario when setting this up. Removed. Happy to create a PR in the future to re-introduce it in a cleaner way that prevents this issue.

1. Yes that should be possible, the schema already allows for one to many associations, I don't see any database schema issues. Can also propose that in a future PR, definitely useful. Currently we have the issue that a locally registered user can not use an OIDC account with the same email. This is not a limitation of the oidc_accounts table but the existing User schema. Imo this is fine as long as we handle the OIDC association properly in the account settings like you proposed.

2. purely speculative as a nice to have. Removed as per your valid points.

Besides these points I have also just pushed a commit concerning your remaining points.

AlexBocken commented:
Getting lots of unit test fails for the Python 3.9 build, even outside of my added tests. Should this be the case?

DavidMStraub commented:
I had the same problem and solved it by unpinning sentence transformers, it should work again after you've rebased on `master`.

AlexBocken commented:
rebased, thanks for that

AlexBocken commented:
Can confirm that with the rebase the tests seem to run smoothly both for 3.9 and 3.13, at least locally. Let's hope the same also happens once you run them via Action :smile:
